File: modules/shared/src/main/scala/shared/topics/TopicCreation.scala
package shared.topics

import io.simplesource.kafka.spec.TopicSpec

final case class TopicCreation(topicName: String, topicSpec: TopicSpec)

object TopicCreation {
  def apply(topicConfig: TopicConfig)(topicType: String): TopicCreation = {
    val name = topicConfig.namer(topicType)
    val spec = topicConfig.topicSpecs(topicType)
    new TopicCreation(name, spec)
  }

  def withCustomName(topicConfig: TopicConfig, topicType: String)(topicName: String): TopicCreation = {
    val spec = topicConfig.topicSpecs(topicType)
    new TopicCreation(topicName, spec)
  }

  def allTopics(topicConfig: TopicConfig): List[TopicCreation] = {
    topicConfig.topicSpecs.map {
      case (topicBase, config) => {
        val name = topicConfig.namer(topicBase)
        _root_.shared.topics.TopicCreation(name, config)
      }
    }.toList
  }
}

File: java/simplesaga-shared/src/main/java/io/simplesource/saga/shared/topics/TopicCreation.java
package io.simplesource.saga.shared.topics;

import io.simplesource.kafka.spec.TopicSpec;
import lombok.Value;

import java.util.List;
import java.util.stream.Collectors;

@Value
public class TopicCreation {
    public String topicName;
    public TopicSpec topicSpec;

    public static TopicCreation apply(TopicConfig topicConfig, String topicType) {
        String name = topicConfig.namer.apply(topicType);
        TopicSpec spec = topicConfig.topicSpecs.get(topicType);
        return new TopicCreation(name, spec);
    }

    public static TopicCreation withCustomName(TopicConfig topicConfig, String topicType, String topicName) {
        TopicSpec spec = topicConfig.topicSpecs.get(topicType);
        return new TopicCreation(topicName, spec);
    }

    public static List<TopicCreation> allTopics(TopicConfig topicConfig) {
        return topicConfig.topicSpecs.entrySet().stream().map(kv -> {
            String topicBase = kv.getKey();
            String name = topicConfig.namer.apply(topicBase);
            return new TopicCreation(name, kv.getValue());
        }).collect(Collectors.toList());
    }
}
